source/fab/tools/compiler.py

```diff
@@ -0,0 +1,301 @@
+                zlib.crc32(str(self.get_version()).encode()))
+
+    def compile_file(self, input_file: Path, output_file: Path,
+                     add_flags: Union[None, List[str]] = None):
+        '''Compiles a file. It will add the flag for compilation-only
+        automatically, as well as the output directives. The current working
+        directory for the command is set to the folder where the source file
+        lives when compile_file is called. This is done to stop the compiler
+        inserting folder information into the mod files, which would cause
+        them to have different checksums depending on where they live.
+
+        :param input_file: the path of the input file.
+        :param outpout_file: the path of the output file.
+        :param add_flags: additional compiler flags.
+        '''
+
+        params: List[Union[Path, str]] = [self._compile_flag]
+        if add_flags:
+            params += add_flags
+
+        params.extend([input_file.name,
+                      self._output_flag, str(output_file)])
+
+        return self.run(cwd=input_file.parent,
+                        additional_parameters=params)
+
+    def check_available(self) -> bool:
+        '''
+        :returns: whether the compiler is available or not. We do
+            this by requesting the compiler version.
+        '''
+        try:
+            version = self.get_version()
+        except RuntimeError:
+            # Compiler does not exist:
+            return False
+
+        # An empty string is returned if some other error occurred when trying
+        # to get the compiler version.
+        return version != ""
+
+    def get_version(self):
+        """
+        Try to get the version of the given compiler.
+        # TODO: why return "" when an error happened?
+        # TODO: we need to properly create integers for compiler versions
+        #       to (later) allow less and greater than comparisons.
+
+        Expects a version in a certain part of the --version output,
+        which must adhere to the n.n.n format, with at least 2 parts.
+
+        :Returns: a version string, e.g '6.10.1', or empty string if
+            a different error happened when trying to get the compiler version.
+
+        :raises RuntimeError: if the compiler was not found.
+        """
+        if self._version:
+            return self._version
+
+        try:
+            res = self.run("--version", capture_output=True)
+        except FileNotFoundError as err:
+            raise RuntimeError(f'Compiler not found: {self.name}') from err
+        except RuntimeError as err:
+            self.logger.warning(f"Error asking for version of compiler "
+                                f"'{self.name}': {err}")
+            return ''
+
+        # Pull the version string from the command output.
+        # All the versions of gfortran and ifort we've tried follow the
+        # same pattern, it's after a ")".
```

Comment:
And that's the reason why there is no Cray compiler defined atm :)  My idea is that we properly handle version (i.e. breaking them into integer-tuples, storing them, and using them in version-specific tests). At that time I would add Cray plus a custom function to query and extract the version number.

Reply:
Great. Please raise a ticket with the details so it doesn't get lost.

Reply:
#318. Note that it will be a while before we get to that, since we have to support compiler-version-specific compiler flags, so changes to the flag handling need to come first (or well, all together ... or maybe I split this in the end ... anyway, work-in-progress).